Added GetAssemblyName API for AssemblyDefinition and AssemblyReference

maryamariyan opened the pull request:
Also added GetAssemblyName() APIs on AssemblyReference and
AssemblyDefinition

This is my initial pull request. 
I am looking to get feedback.

fixes #13295 

cc: @safern @danmosemsft @tmat @nguerrera

dnfclas commented:
@__maryamariyan__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/AssemblyDefinition.cs:
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Diagnostics;
+using System.Runtime;
```
I don't think this is necessary.

maryamariyan replied:
removed in the new changeset

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/AssemblyReference.cs:
```diff
@@ -4,6 +4,7 @@
 
 using System.Diagnostics;
 using System.Reflection.Metadata.Ecma335;
+using System.Runtime;
```
Neither do this one.

maryamariyan replied:
removed this as well

safern commented on src/System.Reflection.Metadata/tests/Utilities/StringUtilsTests.cs:
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using Xunit;
+using System.Reflection.Metadata;
```
This change is not necessary either.

maryamariyan replied:
removed this as well

dnfclas commented:
__@maryamariyan__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs:
```diff
@@ -1117,6 +1117,35 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
+
+            assemblyName.SetPublicKey(publicKeyOrToken);
+            //assemblyName.Flags = AssemblyTable.GetFlags();
+            //assemblyName.HashAlgorithm = AssemblyTable.GetHashAlgorithm();
```
Talked to @maryamariyan and she will update this to actually calculate the flag values and assign them correctly.

maryamariyan replied:
thanks @safern could you please now take a look at the new changeset and let me know what you think?

danmosemsft commented on src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/StringUtils.cs:
```diff
@@ -1,6 +1,7 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
+using System.Collections.Immutable;
```
revert?

maryamariyan replied:
thanks. I reverted this in the new changeset

danmosemsft commented:
Looks good so far, please include tests in the same PR.

maryamariyan commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs:
```diff
@@ -1117,6 +1117,108 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
+            return assemblyName;
+        }
+
+        private void SetAssemblyNameFlags(AssemblyName assemblyName)
```
@tmat Would it make sense to move this private method (SetAssemblyNameFlags) to AssemblyName and call it SetFlags() ? or is this ok here as a private method?

maryamariyan replied:
I guess another option is to move them to a AssemblyNameUtils helper class instead

maryamariyan commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs:
```diff
@@ -1117,6 +1117,108 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
+            bool hasWindowsRuntime = (flags & AssemblyFlags.WindowsRuntime) != 0;
+        }
+
+        private void SetAssemblyNameHashAlgorithm(AssemblyName assemblyName)
```
same comment here.

@tmat Would it make sense to move this private method (SetAssemblyNameHashAlgorithm) to AssemblyName and call it SetHashAlrgorithm() ? or is this ok here as a private method?

maryamariyan replied:
same comment here.

optionally perhaps I could move this to a AssemblyNameUtils helper class instead

maryamariyan commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs:
```diff
@@ -1117,6 +1117,108 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
+
+        private void SetJitCompileOptimizerEnabled(AssemblyName assemblyName, AssemblyFlags flags)
+        {
+            // notice we are setting EnableJITcompileOptimizer from DisableJitCompileOptimizer. Logic is flipped
```
here I am setting EnableJITcompileOptimizer flag from the value of AssemblyFlags.DisableJitCompileOptimizer

notice how on line 1212 the condition check is "!jitCompileTrackingDisabled"

maryamariyan commented:
thanks @danmosemsft. I'll start adding the tests to this PR now.

maryamariyan commented on src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/AssemblyName.cs:
```diff
@@ -0,0 +1,82 @@
+
+            SetJitCompileOptimizerEnabled(assemblyName, flags);
+
+            // the following two flags have no match in AssemblyNameFlags
```
There is no mapping for the two below flags. 
This means, if these two flags happen to be set, we won't be able to set any value in AssemblyName for them. Is this OK?

safern replied:
Actually you could just ignore them and delete this two bool properties :) we should just fill as much information as possible in AssemblyName.

MichalStrehovsky replied:
> Actually you could just ignore them 

@maryamariyan @safern This should not be ignored. It should be set on `AssemblyName.ContentType`. The `AssemblyFlags` to `AssemblyContentType` conversion is done using a magic formula, like so: https://github.com/dotnet/corert/blob/6468f5fc5da3d39ad0515a8821f246e988aca662/src/Common/src/TypeSystem/Ecma/EcmaModule.cs#L527

safern replied:
Thanks @MichalStrehovsky didn't know about that. Thanks for pointing it out :)

maryamariyan commented on src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/AssemblyName.cs:
```diff
@@ -0,0 +1,82 @@
+
+        private static void SetJitCompileOptimizerEnabled(AssemblyName assemblyName, AssemblyFlags flags)
+        {
+            // notice we are setting EnableJITcompileOptimizer from DisableJitCompileOptimizer. Logic is flipped
```
@safern notice how I am setting EnableJITcompileOptimizer below. I use AssemblyFlags.DisableJitCompilerOptimizer for this.

notice how on line 72 the if condition is (!jitCompileTrackingDisabled)

safern commented on src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:
```diff
@@ -1608,6 +1610,7 @@ public sealed partial class MetadataReader
         public System.Reflection.Metadata.AssemblyDefinition GetAssemblyDefinition() { throw null; }
         public System.Reflection.Metadata.AssemblyFile GetAssemblyFile(System.Reflection.Metadata.AssemblyFileHandle handle) { throw null; }
         public System.Reflection.Metadata.AssemblyReference GetAssemblyReference(System.Reflection.Metadata.AssemblyReferenceHandle handle) { throw null; }
+        public System.Reflection.AssemblyName GetAssemblyName() { throw null; }
```
Do we want to expose this API as public or make it instead internal? As the API proposal was to add this methods in AssemblyDefinition and AssemblyReference.

nguerrera replied:
We should not expose it here too. I think it only clutters the API. `GetAssemblyDefinition().GetAssemblyName() ` is fine.

maryamariyan replied:
Will be removing this exposed API in the next changelist

safern commented on src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:
```diff
@@ -99,6 +99,7 @@ public partial struct AssemblyDefinition
         public System.Reflection.Metadata.StringHandle Name { get { throw null; } }
         public System.Reflection.Metadata.BlobHandle PublicKey { get { throw null; } }
         public System.Version Version { get { throw null; } }
+        public System.Reflection.AssemblyName AssemblyName { get { throw null; } }
```
This should instead be a method and not a property as that was the approved API. 

```cs
public System.Reflection.AssemblyName GetAssemblyName()
```

safern commented on src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:
```diff
@@ -164,6 +165,7 @@ public partial struct AssemblyReference
         public System.Reflection.Metadata.StringHandle Name { get { throw null; } }
         public System.Reflection.Metadata.BlobHandle PublicKeyOrToken { get { throw null; } }
         public System.Version Version { get { throw null; } }
+        public System.Reflection.AssemblyName AssemblyName { get { throw null; } }
```
This should instead be a method and not a property as that was the approved API. 

```cs
public System.Reflection.AssemblyName GetAssemblyName()
```

nguerrera replied:
Yes. The rationale here is that being consumed by very performance sensitive callers, this library is very careful to designate things which take time and allocate as methods, not properties.

maryamariyan replied:
I see. Thanks @nguerrera for your explanation

danmosemsft commented:
Failures above are due to UnloadingEventMustHappenBeforeProcessExitEvent which has been fixed

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/AssemblyName.cs:
```diff
@@ -0,0 +1,82 @@
+                [AssemblyHashAlgorithm.Sha256] = Configuration.Assemblies.AssemblyHashAlgorithm.SHA256,
+                [AssemblyHashAlgorithm.Sha384] = Configuration.Assemblies.AssemblyHashAlgorithm.SHA384,
+                [AssemblyHashAlgorithm.Sha512] = Configuration.Assemblies.AssemblyHashAlgorithm.SHA512
+            };
```
This is going to allocate a new dictionary on every call. I think a `switch` would do fine for this.

tmat replied:
Actually, I'd just use a cast, otherwise we will need to update the code every time a new algorithm is added. Perhaps also add a test that validates that these enums have the same items.

danmosemsft commented:
That means that the net native one is not seeing the extension methods. I can't immediately see why. Maybe Santi can.

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/AssemblyName.cs:
```diff
@@ -0,0 +1,82 @@
+
+        internal static void SetFlags(this AssemblyName assemblyName, AssemblyFlags flags)
+        {
+            var flagMappings = new Dictionary<AssemblyFlags, AssemblyNameFlags>()
```
This will also allocate a Dictionary on every call. It looks prettier but is not as efficient as looking at if the flag is set through an & operation and then set it if so. i.e

```cs
if ((AssemblyFlags.PublicKey & flags) != 0)
     assemblyName.Flags |= AssemblyNameFlags.PublickKey
```

But if we add more flags later we would need to change this code, so perhaps using a cast is better, just as suggested here: https://github.com/dotnet/corefx/pull/23396/files#r134070466

safern commented:
I took a look at the errors in UWP NETNative, the reason why this is happening is because System.Reflection.Metadata is compiled for netcoreapp only but we have a uapaot version which is: `System.Private.Reflection.Metadata.Ecma335.csproj`. That csproj is built for uapaot only and it imports all of the System.Reflection.Metadata files, which includes the MetadataReader. 

Since you added some internal helper extensions methods to the MetadataReader in `src/System/Reflection/Internal/Utilities/AssemblyName.cs` and using them in the MetadataReader you will need to include this file into `System.Private.Reflection.Metadata.Ecma335.csproj` as well so that it can find those extensions methods.

nguerrera commented:
Tagging @davidwrighton on the last point, it might be better to leave this dependency out of the private version. I'm not sure.

maryamariyan commented:
I updated the PR. Updated the changes by applying the code review comments above. 

@tmat @NickCraver @danmosemsft @safern see if you prefer the newly updated changes of GetAssemblyName() in MetadataReader.cs now.

Next, I'm going to focus on fixing the failures and adding unit tests.
Also, I will look closely at your comments and will apply them accordingly.

Thanks

NickCraver commented:
@maryamariyan I've got a few uses for this, but nothing immediate - possibly a ping meant for @nguerrera?

maryamariyan commented:
yes @NickCraver thanks :) typo. I meant to ping @nguerrera above,

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs:
```diff
@@ -1117,6 +1117,60 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
+            if ((flags & AssemblyFlags.Retargetable) != 0)
+                assemblyName.Flags |= AssemblyNameFlags.Retargetable;
+            else
+                assemblyName.Flags &= AssemblyNameFlags.Retargetable;
```
This doesn't look right. Won't it clear the other flags when the else branch is taken?

nguerrera replied:
I think you meant `&= ~flag`, but see note about just casting below.

maryamariyan replied:
oh no. You're right. will fix this in next changeset

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs:
```diff
@@ -1117,6 +1117,60 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
+            if ((flags & AssemblyFlags.DisableJitCompileOptimizer) == 0)
+                assemblyName.Flags |= AssemblyNameFlags.EnableJITcompileOptimizer;
+            else
+                assemblyName.Flags &= AssemblyNameFlags.EnableJITcompileOptimizer;
```
The reason for the flip here is that AssemblyNameFlags has a hilarious naming bug. 

See https://msdn.microsoft.com/en-us/library/system.reflection.assemblynameflags(v=vs.110).aspx (Emphasis mine.):

> EnableJITcompileOptimizer
>
>Specifies that just-in-time (JIT) compiler optimization is disabled for the assembly. *This is the exact opposite of the meaning that is suggested by the member name.*

So, in fact, the true meaning of the bit is the same between both enumerations. Furthermore, both use the numeric value specified by ECMA 335 and you can therefore just cast here too.

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/AssemblyReference.cs:
```diff
@@ -122,6 +122,11 @@ public BlobHandle HashValue
+        public AssemblyName GetAssemblyName()
+        {
+            return _reader.GetAssemblyName();
+        }
```
This doesn't seem right. You need to fetch the name details for the current AssemblyReference using the row id and/or other members of this type. Once you do that, I don't think there will be much to share between def and ref and you won't even need the internal method on MetadataReader.

nguerrera replied:
Actually, you probably still want a shared static internal helper that takes String and blob handles for name, culture, token, etc. and produces an AssemblyName.

The key point is that each AssemblyReference has a distinct AssemblyName and they're all distinct from the AssemblyName of the AssemblyDefinition.

nguerrera commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyReferenceTests.cs:
```diff
@@ -0,0 +1,29 @@
+        public void DefaultAssemblyNameFlagIsNone()
+        {
+            AssemblyName assemblyName = new AssemblyName();
+
```
This is just testing the AssemblyName default constructor which doesn't live in this assembly. Was this just to confirm its behavior? We shouldn't leave this out of place test here.

maryamariyan replied:
yes, just a placeholder to add other tests

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+    {
+        internal AssemblyName GetAssemblyName(StringHandle nameHandle, Version version, StringHandle cultureHandle, BlobHandle publicKeyHandle, AssemblyHashAlgorithm assemblyHashAlgorithm, AssemblyFlags flags)
+        {
+            string name = (!nameHandle.IsNil) ? GetString(nameHandle) : null;
```
Should we handle this and throw an exception? Maybe an `InvalidDataException` or something; because passing null to `AssemblyName..ctor(string)` will throw a `NullReferenceException`

cc: @nguerrera

nguerrera replied:
Please test, but GetString(nil) should throw BadImageFormatException already, which would be the right exception here. So I think you can simply avoid getting in the away of that and use nameHandle unconditionally.

safern replied:
Oh ok. So doing `string name = GetString(nameHandle);` should be enough.

nguerrera replied:
Yes, but please check my memory on that with some testing.

maryamariyan replied:
@nguerrera I wanted to add a test case to check for exception when nameHandle is null. But when I trace the code back I see there is no way to get a null nameHandle or null assemblyName

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+                CultureName = cultureName,
+                HashAlgorithm = hashAlgorithm
+            };
+            assemblyName.SetPublicKey(publicKeyOrToken);
```
Nit => add extra line at top of this one and delete extra line at bottom.

```cs
...
};

assemblyName.SetPublicKey(publicKeyOrToken);
assemblyName.Flags = GetFlagsFromAssemblyFlags(flags);
assemblyName.ContentType = GetContentTypeFromAssemblyFlags(flags);
return assemblyName;
```

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+            return assemblyName;
+        }
+
+        private AssemblyNameFlags GetFlagsFromAssemblyFlags(AssemblyFlags flags)
```
Nit => maybe rename to `GetAssemblyNameFlags`

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+
+        private AssemblyNameFlags GetFlagsFromAssemblyFlags(AssemblyFlags flags)
+        {
+            AssemblyNameFlags assemblyNameFlags = AssemblyNameFlags.None;
```
I don't think this is necessary as since this is a `[Flag]` enum by default the bits will be all set to 0, so it will already have the `.None` flag.

safern replied:
I think you can only do the declaration; no need of assigning, the runtime default value will be an empty flag (all bits turned to 0)

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/AssemblyDefinition.netstandard2.0.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public partial struct AssemblyDefinition
+    {
+
```
Nit => extra line.

safern commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/AssemblyReference.netstandard2.0.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public partial struct AssemblyReference
+    {
+
```
Nit => extra line.

safern commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyDefinitionTests.cs:
```diff
@@ -0,0 +1,43 @@
+            Assert.Equal((uint)0, (uint)assemblyName.Flags); // AssemblyFlags
+
+            // Locale
+            Assert.Equal(null, assemblyName.CultureName);
```
You can do `Assert.Null(assemblyName.CultureName):`

safern commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyDefinitionTests.cs:
```diff
@@ -0,0 +1,43 @@
+            Assert.Equal(null, assemblyName.CultureName);
+
+            // PublicKey
+            Assert.Equal(null, assemblyName.GetPublicKey());
```
Ditto.

safern commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyReferenceTests.cs:
```diff
@@ -0,0 +1,115 @@
+            Assert.Equal(refVersion, assemblyName.Version);
+
+            // Culture
+            Assert.Equal(null, assemblyName.CultureName);
```
`Assembly.Null(assemblyName.CultureName);`

safern commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyReferenceTests.cs:
```diff
@@ -0,0 +1,115 @@
+            Assert.Equal(refPublicKey, assemblyName.GetPublicKey());
+            
+            // HashAlgorithm
+            var refHashAlgorithm = Configuration.Assemblies.AssemblyHashAlgorithm.None;
```
No need to create the extra variable. You could just do: `Assert.Equal(Configuration.Assemblies.AssemblyHashAlgorithm.None, assemblyName.HashAlgorithm);`

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+            if ((flags & AssemblyFlags.PublicKey) != 0)
+                assemblyNameFlags |= AssemblyNameFlags.PublicKey;
+            else
+                assemblyNameFlags &= ~AssemblyNameFlags.PublicKey;
```
Aren't these else blocks are unnecessary? You start with no flags and add to that. I don't see how you ever need to clear flags.

maryamariyan replied:
you're right. they're not necessary. will update in a few mins

safern replied:
+1

safern commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyReferenceTests.cs:
```diff
@@ -0,0 +1,115 @@
+                Assert.Equal(refVersion, assemblyName.Version);
+
+                // Culture
+                Assert.Equal(null, assemblyName.CultureName);
```
`Assert.Null(assemblyName.CultureName);`

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+                HashAlgorithm = hashAlgorithm
+            };
+
+            assemblyName.SetPublicKey(publicKeyOrToken);
```
I think you need to check if the PublicKey flag is set and only do this in that case and otherwise call SetPublicKeyToken

maryamariyan replied:
@nguerrera  I changed it to 



            bool hasPublicKey = (flags & AssemblyFlags.PublicKey) != 0;
            if (hasPublicKey)
            {
                assemblyName.SetPublicKey(publicKeyOrToken);
            }
            else
            {
                assemblyName.SetPublicKeyToken(publicKeyOrToken);
            }

in my local branch, (haven't pushed it up yet).
but if would fail my test. am I missing something?

MichalStrehovsky replied:
Your tests are wrong too :). You need to change them to check `GetPublicKeyToken` in a couple places.

safern commented on src/System.Reflection.Metadata/src/System.Reflection.Metadata.csproj:
```diff
@@ -73,6 +73,8 @@
     <Compile Include="System\Reflection\Metadata\MetadataReaderProvider.cs" />
     <Compile Include="System\Reflection\Metadata\BlobContentId.cs" />
     <Compile Include="System\Reflection\Metadata\Signatures\PrimitiveSerializationTypeCode.cs" />
+    <Compile Include="System\Reflection\Metadata\TypeSystem\AssemblyDefinition.netstandard2.0.cs" Condition="'$(TargetGroup)' != 'netstandard1.1'" />
```
@nguerrera are you okay with this naming and following the TargetGroup approach instead of the `#if !NETSTANDARD1.1` -- I personally like the TargetGroup - partial classes approach but this is your area so your call. 

Also I'm not sure if it should be named `AssemblyDefinition.netstandard2.0.cs` or `AssemblyDefinition.netstandard.cs` since this will be available on NS >= 2.0 not only on 2.0

safern replied:
Now that I think if we keep the pattern of the target group file name it should be `netstandard2.1` as we already shipped 2.0.

@danmosemsft @weshaggard any thoughts on this?

danmosemsft commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+
+        private AssemblyContentType GetContentTypeFromAssemblyFlags(AssemblyFlags flags)
+        {
+            return (AssemblyContentType)(((int)flags & 0x0E00) >> 9);
```
Can you use AssemblyFlags.ContentTypeMask instead of magic number 0x0E00

danmosemsft commented:
@dotnet-bot test code coverage please

danmosemsft commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyDefinitionTests.cs:
```diff
@@ -0,0 +1,120 @@
+            // PublicKey
+            Assert.Null(assemblyName.GetPublicKey());
+            Assert.Null(assemblyName.GetPublicKeyToken());
+
```
nit: extra line

BTW, many poeple would break these into separate tests, doing the setup repeatedly with a utility function. It's fine like this too though.

maryamariyan replied:
I see. will do

danmosemsft commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.0.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            string name = GetString(nameHandle);
+            string cultureName = (!cultureHandle.IsNil) ? GetString(cultureHandle) : null;
+            var hashAlgorithm = (Configuration.Assemblies.AssemblyHashAlgorithm)assemblyHashAlgorithm;
```
Aside: lame that there are two enums with the same name and (fortunately) the same values

danmosemsft commented:
@dotnet-bot test code coverage please (don't wait on this -- experimenting)

tmat commented on src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:
```diff
@@ -99,6 +99,9 @@ public partial struct AssemblyDefinition
         public System.Reflection.Metadata.BlobHandle PublicKey { get { throw null; } }
         public System.Version Version { get { throw null; } }
+#if !NETSTANDARD11
+        public System.Reflection.AssemblyName GetAssemblyName() { throw null; }
```
>public System.Reflection.AssemblyName GetAssemblyName() { throw null; } [](start = 8, length = 71)

Why are these APIs netstandard >1.1 only?

maryamariyan replied:
Because System.Reflection.AssemblyName was not introduced in netstandard1.1 yet

safern replied:
This was my mistake to help her fix the AllConfigurations build errors, but from what I just checked they are available but in a reference that was not included when building for ns1.1. I will work with her to get this fixed.

maryamariyan replied:
@safern, I had an offline conversation with @tmat regarding our recent experimentation when we try to add GetAssemblyName API in netstandard1.1.

It seems that it should be fine to just add this API to NS2.0

tmat replied:
To clarify, we could implement a light-up but it's non-trivial. I'll file a separate work item for that.

tmat replied:
https://github.com/dotnet/corefx/issues/23487

safern commented:
@davidwrighton do you want this to be included into `System.Private.Reflection.Metadata.Ecma335.csproj`? So that it is available in uap?

tmat commented:
@safern Let's not include it in the SRM private. We want that to be minimal.

maryamariyan commented:
@tmat @nguerrera this PR is ready for code review

tmat commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyDefinitionTests.cs:
```diff
@@ -0,0 +1,76 @@
+        [Fact]
+        public void ValidateAssemblyNameForAssemblyDefinition()
+        {
+            var assemblyItems = new[] {
```
>{ [](start = 38, length = 1)

Nit: brace on a new line.

tmat commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyDefinitionTests.cs:
```diff
@@ -0,0 +1,76 @@
+                new { Assembly = WinRT.Lib, Name = "Lib", Version = new Version(1, 0, 0, 0), ContentType = AssemblyContentType.WindowsRuntime },
+                new { Assembly = NetModule.AppCS, Name = "AppCS", Version = new Version(1, 2, 3, 4), ContentType = AssemblyContentType.Default },
+                new { Assembly = Namespace.NamespaceTests, Name = "NamespaceTests", Version = new Version(0, 0, 0, 0), ContentType = AssemblyContentType.Default },
+                new { Assembly = PortablePdbs.DocumentsDll, Name = "Documents", Version = new Version(0, 0, 0, 0), ContentType = AssemblyContentType.Default }
```
Missing a test for assembly name with a culture set.

tmat commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard2.1.cs:
```diff
@@ -0,0 +1,62 @@
+{
+    public sealed partial class MetadataReader
+    {
+        internal AssemblyName GetAssemblyName(StringHandle nameHandle, Version version, StringHandle cultureHandle, BlobHandle publicKeyHandle, AssemblyHashAlgorithm assemblyHashAlgorithm, AssemblyFlags flags)
```
>publicKeyHandle [](start = 127, length = 15)

Nit: i'd name this parameter "publicKeyOrTokenHandle" as it can be either

safern commented on src/System.Reflection.Metadata/src/System.Reflection.Metadata.csproj:
```diff
@@ -73,6 +73,8 @@
     <Compile Include="System\Reflection\Metadata\MetadataReaderProvider.cs" />
     <Compile Include="System\Reflection\Metadata\BlobContentId.cs" />
     <Compile Include="System\Reflection\Metadata\Signatures\PrimitiveSerializationTypeCode.cs" />
+    <Compile Include="System\Reflection\Metadata\TypeSystem\AssemblyDefinition.netstandard2.1.cs" Condition="'$(TargetGroup)' != 'netstandard1.1'" />
```
I would probably name this as netstandard2.0 or just netstandard (versionless) as that is the pattern we've been following when it is available in latest version.

tmat commented on src/System.Reflection.Metadata/tests/Metadata/TypeSystem/AssemblyDefinitionTests.cs:
```diff
@@ -0,0 +1,148 @@
+            bool hasDefaultCulture = true, 
+            bool hasDefaultHashAlgorithm = true, 
+            bool hasDefaultPublicKey = true, 
+            bool hasDefaultFlags = true)
```
These parameters seem a bit too complicated. 
Do we actually need this function? Why not just compare the actual values against the expected values (default or not)?

maryamariyan replied:
done
